File: keychain-sample-chat-iphone/KeychainChat/protocols/ChatRepository.swift
//
//  DBService.swift
//  KeychainChat
//
//  Created by Robert Ellis on 2022/06/10.
//

import Foundation
import SwiftUI

protocol ChatRepository {
    func createOrOpenDB() throws
    
    func getPlatformUsers(filterBy: [String: Facade], completion: @escaping ([String: User]) -> Void)
    
    func getPlatformUser(recordId: String, completion: @escaping (User?) -> Void)
    
    func getPlatformUser(firstName: String, lastName: String, completion: @escaping (User?) -> Void)
    
    func getPlatformUser(uri: String, completion: @escaping (User?) -> Void)
    
    func saveUserProfile(firstName: String, lastName: String, status: Int, uri: String?, image: UIImage?, completion: @escaping (String?) -> Void)
    
    func updateUserProfile(uri: String, firstName: String, lastName: String?, completion: @escaping (String?) -> Void)
    
    func updateUserProfile(firstName: String, lastName: String, status: Int, uri: String?, completion: @escaping (String?) -> Void)
    
    func deleteUser(uri: String, completion: @escaping (String?) -> Void)
    
    func getAllChats(senderId: String, completion: @escaping ([Chat]) -> Void)
    
    func getChat(senderId: String, receiverId: String, completion: @escaping (Chat?) -> Void)
    
    func getAllMessages(chat: Chat, completion: @escaping ([ChatMessage]) -> Void)
    
    func saveChat(chat: Chat, completion: @escaping (String?) -> Void)
    
    func updateChat(id: String, lastMessage: String)
    
    func isMessageExisting(recordId: String) -> Bool?
    
    func getMessage(recordId: String, completion: @escaping (ChatMessage?) -> Void)

    func saveMessage(chatMsgId: String?,
                     senderId: String,
                     msg: String,
                     direction: ChatDirection,
                     chat: Chat, completion: @escaping (String?) -> Void)
    
    func savePhotoMessage(senderUri: String, image: UIImage, chat: Chat)
}

File: keychain-sample-chat-android/app/src/main/java/io/keychain/chat/interfaces/ChatRepository.java
package io.keychain.chat.interfaces;

import android.graphics.Bitmap;
import android.media.Image;

import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import io.keychain.chat.models.chat.Chat;
import io.keychain.chat.models.chat.ChatDirection;
import io.keychain.chat.models.chat.ChatMessage;
import io.keychain.chat.models.chat.User;

public interface ChatRepository {

    Optional<Map<String, User>> getPlatformUsers(Set<String> filterBy);

    Optional<User> getPlatformUser(String recordId);

    Optional<List<User>> getPlatformUser(String firstName, String lastName);

    Optional<User> getPlatformUserByUri(String uri);

    // Returns the record id
    Optional<String> saveUserProfile(String firstName, String lastName, int status, int source, String uri, Bitmap image);

    boolean updateUserProfile(String firstName, String lastName, int status, String uri);

    Optional<List<Chat>> getAllChats(String senderId, Set<String> personaIds);

    Optional<Chat> getChat(String senderId, String receiverId);

    Optional<List<ChatMessage>> getAllMessages(Chat chat);

    Optional<List<ChatMessage>> getAllMessages(String uri);

    // Returns the record id
    Optional<String> saveChat(Chat chat);

    boolean updateChat(String id, String lastMessage);

    Optional<ChatMessage> getMessage(String recordId);

    // Returns the record id
    Optional<String> saveMessage(String chatId, String senderId, String msg, ChatDirection direction, Chat chat);

    // Returns the record id
    Optional<String> savePhotoMessage(String senderUri, Image image, Chat chat);
}
